school_lunch/controllers/website_sale.py

```diff
@@ -36,33 +36,37 @@ def confirm_order(self, **post):
-        if promo:
-            pricelist_sudo = request.env["product.pricelist"].sudo().search([("code", "=", promo)], limit=1)
-            if not (pricelist_sudo and request.website.is_pricelist_available(pricelist_sudo.id)):
-                return request.redirect("%s?code_not_available=1" % redirect)
 
-            request.session["website_sale_current_pl"] = pricelist_sudo.id
-            # TODO find the best way to create the order with the correct pricelist directly ?
-            # not really necessary, but could avoid one write on SO record
-            order_sudo = request.website.sale_get_order(force_create=True)
-            order_sudo._cart_update_pricelist(pricelist_id=pricelist_sudo.id)
-        else:
-            order_sudo = request.website.sale_get_order()
-            # PATCH START
-            test_mode = getattr(threading.current_thread(), "testing", False) or request.env.registry.in_test_mode()
-            if order_sudo and test_mode:
-                order_sudo._cart_update_pricelist(update_pricelist=True)
-            # PATCH END
-        return request.redirect(redirect)
+@http.route(["/shop/pricelist"], type="http", auth="public", website=True, sitemap=False)
+def pricelist(self, promo, **post):
+    """
+    Standard code reference :
+    https://github.com/odoo/odoo/blob/3383d5bd68bfc13b7881f72e5adbb7c27a6df30e/addons/website_sale/controllers/main.py
+    In the original file, `pricelist` is taken from l.750 to l.769
+    In the standard code, the Sale Order's pricelist is updated when an empty promo code is used.
+    This is unwanted as the SO has Sales Order Lines with different pricelists,
+    and should not be overwritten by the SO's pricelist.
+    """
+    redirect = post.get("r", "/shop/cart")
+    # empty promo code is used to reset/remove pricelist (see `sale_get_order()`)
+    if promo:
+        pricelist_sudo = request.env["product.pricelist"].sudo().search([("code", "=", promo)], limit=1)
+        if not (pricelist_sudo and request.website.is_pricelist_available(pricelist_sudo.id)):
+            return request.redirect("%s?code_not_available=1" % redirect)
+
+        request.session["website_sale_current_pl"] = pricelist_sudo.id
+        # TODO find the best way to create the order with the correct pricelist directly ?
+        # not really necessary, but could avoid one write on SO record
+        order_sudo = request.website.sale_get_order(force_create=True)
+        order_sudo._cart_update_pricelist(pricelist_id=pricelist_sudo.id)
+    else:
+        order_sudo = request.website.sale_get_order()
+        # PATCH START
+        test_mode = getattr(threading.current_thread(), "testing", False) or request.env.registry.in_test_mode()
+        if order_sudo and test_mode:
+            order_sudo._cart_update_pricelist(update_pricelist=True)
+        # PATCH END
+    return request.redirect(redirect)
+
+
+main.WebsiteSale.pricelist = pricelist
```

Comment:
I believe that this patch can be simplified to avoid breaking it in the future if the original method changes and also this suggestion will lead to have less cloc (less maintainable code, less money for the client) and this is by using context variables.
```suggestion
    @http.route()
    def pricelist(self, promo, **post):
        test_mode = getattr(threading.current_thread(), "testing", False) or request.env.registry.in_test_mode()
        if test_mode:
            request.update_context(school_not_update_pricelist=True)
        return super().pricelist(promo, **post)
```
and in `sale.order` model:
```python
    def _cart_update_pricelist(self, pricelist_id=None, update_pricelist=False):
        if self.env.context.get('school_not_update_pricelist'):
            # maybe you just need to set update_pricelist = False
            return
        return super()._cart_update_pricelist(pricelist_id, update_pricelist)
```
can you please test it, and you can do the same for `confirm_order()` as well

Reply:
Hi @Bechir-bbr , thanks for your suggestion, context parameters seem very useful in this case! I will try taking them into account when patching code.

Yet I have tried your code on my local db: while the context is updated through the controller's `pricelist` function, when the context is evaluated in `sale.order::_cart_update_pricelist` via `self.env.context.get('school_not_update_pricelist')`, the result is always `False`

Reply:
you're right, the context should be passed when calling `_cart_update_pricelist()` , humm I think we can't really apply it in this use case